File: leetcode/201-300/276-paint-fence/bench/paint_enum.c
/* LeetCode 276 bench mirror — brute-force enumeration, C.
 *
 * Same algorithm as paint_enum.kara: split on the first two posts, odometer the
 * suffix, check the whole array for three-in-a-row. Sequential: the 16 prefixes
 * run in a plain loop. */
#include <stdio.h>
#include <stdint.h>

#define N 13
#define K 4

static int64_t count_prefix(int p0, int p1) {
    int c[N] = {0};
    c[0] = p0;
    c[1] = p1;
    int64_t count = 0;
    for (;;) {
        int ok = 1;
        for (int i = 2; i < N; i++)
            if (c[i] == c[i - 1] && c[i - 1] == c[i - 2]) ok = 0;
        if (ok) count++;
        int p = N - 1;
        while (p >= 2 && c[p] == K - 1) { c[p] = 0; p--; }
        if (p < 2) break;
        c[p]++;
    }
    return count;
}

int main(void) {
    int64_t total = 0;
    for (int pre = 0; pre < K * K; pre++)
        total += count_prefix(pre / K, pre % K);
    printf("%lld\n", (long long)total);
    return 0;
}

File: leetcode/201-300/276-paint-fence/bench/paint_enum.rs
// LeetCode 276 bench mirror — brute-force enumeration, Rust.
// Same algorithm as paint_enum.kara; sequential over the 16 prefixes.

const N: usize = 13;
const K: i64 = 4;

fn count_prefix(p0: i64, p1: i64) -> i64 {
    let mut c = [0i64; N];
    c[0] = p0;
    c[1] = p1;
    let mut count = 0i64;
    loop {
        let mut ok = true;
        for i in 2..N {
            if c[i] == c[i - 1] && c[i - 1] == c[i - 2] {
                ok = false;
            }
        }
        if ok {
            count += 1;
        }
        let mut p = N as i64 - 1;
        while p >= 2 && c[p as usize] == K - 1 {
            c[p as usize] = 0;
            p -= 1;
        }
        if p < 2 {
            break;
        }
        c[p as usize] += 1;
    }
    count
}

fn main() {
    let mut total = 0i64;
    for pre in 0..K * K {
        total += count_prefix(pre / K, pre % K);
    }
    println!("{}", total);
}
